Mathlib/Algebra/Category/BialgebraCat/Basic.lean

```diff
@@ -0,0 +1,199 @@
+    BialgebraCat R where
+  isBialgebra := (inferInstance : Bialgebra R X)
+
+variable {R}
+
+@[simp]
+lemma of_comul {X : Type v} [Ring X] [Bialgebra R X] :
+    Coalgebra.comul (A := of R X) = Coalgebra.comul (R := R) (A := X) := rfl
+
+@[simp]
+lemma of_counit {X : Type v} [Ring X] [Bialgebra R X] :
+    Coalgebra.counit (A := of R X) = Coalgebra.counit (R := R) (A := X) := rfl
+
+/-- A type alias for `BialgHom` to avoid confusion between the categorical and
+algebraic spellings of composition. -/
+@[ext]
+structure Hom (V W : BialgebraCat.{v} R) :=
+  /-- The underlying `BialgHom` -/
+  toBialgHom : V →ₐc[R] W
+
+lemma Hom.toBialgHom_injective (V W : BialgebraCat.{v} R) :
+    Function.Injective (Hom.toBialgHom : Hom V W → _) :=
+  fun ⟨f⟩ ⟨g⟩ _ => by congr
+
+instance category : Category (BialgebraCat.{v} R) where
+  Hom X Y := Hom X Y
+  id X := ⟨BialgHom.id R X⟩
+  comp f g := ⟨BialgHom.comp g.toBialgHom f.toBialgHom⟩
+
+-- TODO: if `Quiver.Hom` and the instance above were `reducible`, this wouldn't be needed.
+@[ext]
+lemma hom_ext {X Y : BialgebraCat.{v} R} (f g : X ⟶ Y) (h : f.toBialgHom = g.toBialgHom) :
+    f = g :=
+  Hom.ext _ _ h
+
+/-- Typecheck a `BialgHom` as a morphism in `BialgebraCat R`. -/
+abbrev ofHom {X Y : Type v} [Ring X] [Ring Y]
+    [Bialgebra R X] [Bialgebra R Y] (f : X →ₐc[R] Y) :
+    of R X ⟶ of R Y :=
+  ⟨f⟩
+
+@[simp] theorem toBialgHom_comp {X Y Z : BialgebraCat.{v} R} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    (f ≫ g).toBialgHom = g.toBialgHom.comp f.toBialgHom :=
+  rfl
+
+@[simp] theorem toBialgHom_id {M : BialgebraCat.{v} R} :
+    Hom.toBialgHom (𝟙 M) = BialgHom.id _ _ :=
+  rfl
+
+instance concreteCategory : ConcreteCategory.{v} (BialgebraCat.{v} R) where
+  forget :=
+    { obj := fun M => M
+      map := fun f => f.toBialgHom }
+  forget_faithful :=
+    { map_injective := fun {M N} => DFunLike.coe_injective.comp <| Hom.toBialgHom_injective _ _ }
+
+instance hasForgetToAlgebra : HasForget₂ (BialgebraCat R) (AlgebraCat R) where
+  forget₂ :=
+    { obj := fun X => AlgebraCat.of R X
+      map := fun {X Y} f => (f.toBialgHom : X →ₐ[R] Y) }
+
+@[simp]
+theorem forget₂_algebra_obj (X : BialgebraCat R) :
+    (forget₂ (BialgebraCat R) (AlgebraCat R)).obj X = AlgebraCat.of R X :=
+  rfl
+
+@[simp]
+theorem forget₂_algebra_map (X Y : BialgebraCat R) (f : X ⟶ Y) :
+    (forget₂ (BialgebraCat R) (AlgebraCat R)).map f = (f.toBialgHom : X →ₐ[R] Y) :=
+  rfl
+
+instance hasForgetToCoalgebra : HasForget₂ (BialgebraCat R) (CoalgebraCat R) where
+  forget₂ :=
+    { obj := fun X => CoalgebraCat.of R X
+      map := fun {X Y} f => CoalgebraCat.ofHom f.toBialgHom }
+
+@[simp]
+theorem forget₂_coalgebra_obj (X : BialgebraCat R) :
+    (forget₂ (BialgebraCat R) (CoalgebraCat R)).obj X = CoalgebraCat.of R X :=
+  rfl
+
+@[simp]
+theorem forget₂_coalgebra_map (X Y : BialgebraCat R) (f : X ⟶ Y) :
+    (forget₂ (BialgebraCat R) (CoalgebraCat R)).map f = CoalgebraCat.ofHom f.toBialgHom :=
+  rfl
+
+end BialgebraCat
```

Comment:
You now deviate from `QuadraticModuleCat.lean`, defining `toIso` rather than `ofIso` in a different namespace, but to be honest I'm more confused about the naming convention in `QuadraticModuleCat` than what's going on here...

Reply:
Yeah, with this I decided to be more like `AlgEquiv.toAlgebraIso`. I guess maybe it should be `BialgEquiv.toBialgebraIso`. There's also `BialgEquiv.toHopfAlgebraIso` in a later PR I think.